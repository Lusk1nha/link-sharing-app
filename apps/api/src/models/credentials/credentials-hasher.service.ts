import { Injectable } from "@nestjs/common";
import { comparePassword, hashPassword } from "src/helpers/hash";

@Injectable()
export class CredentialsHasherService {
  async hash(password: string): Promise<string> {
    return hashPassword(password);
  }

  async compare(password: string, hash: string): Promise<boolean> {
    return comparePassword(password, hash);
  }
}
