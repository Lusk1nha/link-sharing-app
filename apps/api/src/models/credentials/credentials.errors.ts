import { HttpStatus } from '@nestjs/common';
import { BaseHttpException } from 'src/common/exceptions/base-expections.common';

export class CredentialAlreadyExistsForUserException extends BaseHttpException {
  constructor() {
    super(
      {
        message: 'Credential already exists for this user.',
        code: 'CREDENTIAL_ALREADY_EXISTS',
        status: HttpStatus.CONFLICT,
      },
      HttpStatus.CONFLICT,
      'CREDENTIAL_ALREADY_EXISTS',
    );
  }
}

export class CredentialNotFoundException extends BaseHttpException {
  constructor() {
    super(
      {
        message: 'Credential not found.',
        code: 'CREDENTIAL_NOT_FOUND',
        status: HttpStatus.NOT_FOUND,
      },
      HttpStatus.NOT_FOUND,
      'CREDENTIAL_NOT_FOUND',
    );
  }
}
