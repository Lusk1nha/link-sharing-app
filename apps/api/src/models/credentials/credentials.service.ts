import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from 'src/common/database/database.service';
import { UUID } from 'src/common/entities/uuid/uuid.entity';

import { CredentialEntity } from './domain/credential.entity';
import { CredentialMapper } from './domain/credential.mapper';
import {
  CredentialAlreadyExistsForUserException,
  CredentialNotFoundException,
} from './credentials.errors';
import { PrismaTransaction } from 'src/common/database/__types__/database.types';
import { Credential, Prisma } from '@prisma/client';
import { PrismaBaseService } from 'src/common/database/database-base.service';

@Injectable()
export class CredentialsService extends PrismaBaseService<Credential> {
  protected readonly modelName = 'credential';

  private readonly logger = new Logger(CredentialsService.name);

  constructor(protected readonly prisma: PrismaService) {
    super(prisma);
  }

  /**
   * Retrieves a credential by unique criteria, mapping to domain or returning null.
   */
  private async findUnique(
    where: Prisma.CredentialWhereUniqueInput,
  ): Promise<CredentialEntity | null> {
    const record = await this.prisma.credential.findUnique({ where });
    return record ? CredentialMapper.toDomain(record) : null;
  }

  private async updateUnique(
    credential: CredentialEntity,
    tx?: PrismaTransaction,
  ): Promise<CredentialEntity> {
    const client = this.client(tx);
    const record = await client.credential.update({
      where: { id: credential.id.value },
      data: {
        passwordHash: credential.passwordHash,
      },
    });

    return CredentialMapper.toDomain(record);
  }

  /**
   * Finds a credential by user ID, returning null if not found.
   */
  async findByUserId(userId: UUID): Promise<CredentialEntity | null> {
    const credential = await this.findUnique({ userId: userId.value });

    if (credential) {
      this.logger.debug(`Credential found for userId=${userId.value}`);
    }

    return credential;
  }

  /**
   * Finds a credential by user ID or throws NotFound exception.
   */
  async findByUserIdOrThrow(userId: UUID): Promise<CredentialEntity> {
    const credential = await this.findByUserId(userId);

    if (!credential) {
      this.logger.warn(`No credential for userId=${userId.value}`);
      throw new CredentialNotFoundException();
    }

    return credential;
  }

  /**
   * Creates a credential for a user. Throws if one already exists.
   */
  async createCredential(
    credential: CredentialEntity,
    tx?: PrismaTransaction,
  ): Promise<CredentialEntity> {
    const client = this.client(tx);
    this.logger.log(
      `Attempting to create credential for userId=${credential.userId.value}`,
    );

    try {
      const record = await client.credential.create({
        data: {
          id: credential.id.value,
          user: { connect: { id: credential.userId.value } },
          passwordHash: credential.passwordHash,
        },
      });
      this.logger.log(`Credential created with id=${record.id}`);
      return CredentialMapper.toDomain(record);
    } catch (err) {
      if (
        err instanceof Prisma.PrismaClientKnownRequestError &&
        err.code === 'P2002' &&
        (err.meta?.target as string[]).includes('userId')
      ) {
        this.logger.error(
          `Credential already exists for userId=${credential.userId.value}`,
        );
        throw new CredentialAlreadyExistsForUserException();
      }

      this.logger.error(`Error creating credential: ${err.message}`);
      throw err;
    }
  }

  async updateCredentials(
    userId: UUID,
    hash: string,
  ): Promise<CredentialEntity> {
    this.logger.log(
      `Attempting to update credentials for userId=${userId.value}`,
    );

    return await this.prisma.$transaction(async (tx) => {
      const credential = await this.findByUserIdOrThrow(userId);

      const patched = CredentialEntity.patch(credential, {
        passwordHash: hash,
      });

      return await this.updateUnique(patched, tx);
    });
  }
}
