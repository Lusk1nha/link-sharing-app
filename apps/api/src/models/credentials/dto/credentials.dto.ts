import { EmailAddress } from "src/common/entities/email/email.entity";

export interface CreateCredentialDto {
  email: EmailAddress;
  password: string;
}
