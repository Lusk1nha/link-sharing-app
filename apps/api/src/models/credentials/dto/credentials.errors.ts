import { HttpStatus } from "@nestjs/common";
import { BaseHttpException } from "src/common/exceptions/exceptions.common";

export class NotFoundCredentialsException extends BaseHttpException {
  constructor() {
    super(
      {
        message: "Credentials not found.",
        code: "NOT_FOUND_CREDENTIALS",
      },
      HttpStatus.NOT_FOUND,
      "NOT_FOUND_CREDENTIALS",
    );
  }
}
