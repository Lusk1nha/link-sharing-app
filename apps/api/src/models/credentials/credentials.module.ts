import { Module } from '@nestjs/common';
import { CredentialsService } from './credentials.service';
import { PrismaService } from 'src/common/database/database.service';
import { CredentialsController } from './credentials.controller';

import { PasswordModule } from '../password/password.module';

@Module({
  imports: [PasswordModule],
  controllers: [CredentialsController],
  providers: [CredentialsService, PrismaService],
  exports: [CredentialsService],
})
export class CredentialsModule {}
