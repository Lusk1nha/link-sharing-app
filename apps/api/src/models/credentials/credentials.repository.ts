import { Prisma } from "@prisma/client";
import { PrismaService } from "src/common/prisma/prisma.service";
import { CredentialDto } from "./dto/credentials.model";
import { Injectable } from "@nestjs/common";
import { EmailAddress } from "src/common/entities/email/email.entity";
import { NotFoundCredentialException } from "./dto/credentials.errors";

interface CreateCredentialParams {
  data: Prisma.CredentialCreateInput;
}

interface UpdateCredentialParams {
  data: Prisma.CredentialUpdateInput;
  where: Prisma.CredentialWhereUniqueInput;
}

interface ICredentialsRepository {
  getByEmail(email: EmailAddress): Promise<CredentialDto | null>;
  getByEmailOrThrow(email: EmailAddress): Promise<CredentialDto>;
  create(params: CreateCredentialParams): Promise<CredentialDto>;
  update(params: UpdateCredentialParams): Promise<CredentialDto>;
}

@Injectable()
export class CredentialsRepository implements ICredentialsRepository {
  constructor(private readonly prisma: PrismaService) {}

  async getByEmail(email: EmailAddress): Promise<CredentialDto | null> {
    return this.prisma.credential.findUnique({
      where: {
        email: email.toString(),
      },
    });
  }

  async getByEmailOrThrow(email: EmailAddress): Promise<CredentialDto> {
    const credential = await this.getByEmail(email);

    if (!credential) {
      throw new NotFoundCredentialException();
    }

    return credential;
  }

  async create(params: CreateCredentialParams): Promise<CredentialDto> {
    const { data } = params;

    return this.prisma.credential.create({
      data,
    });
  }

  async update(params: UpdateCredentialParams): Promise<CredentialDto> {
    const { data, where } = params;

    return this.prisma.credential.update({
      data,
      where,
    });
  }
}
