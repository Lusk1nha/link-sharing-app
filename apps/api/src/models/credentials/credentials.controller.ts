import { Body, Controller, Patch } from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import {
  AllowAuthenticated,
  GetAuthUser,
} from 'src/common/auth/auth.decorator';
import { CredentialsService } from './credentials.service';
import { UsersService } from '../users/users.service';
import { JwtStoredPayload } from 'src/common/auth/__types__/auth.types';
import { UpdateCredentialDto } from './dto/update-credential.dto';
import { UUIDFactory } from 'src/common/entities/uuid/uuid.factory';
import { PasswordService } from '../password/password.service';
import { PasswordFactory } from 'src/common/entities/password/password.factory';
import { UpdateCredentialResponseDto } from './dto/update-credential-response.dto';

@Controller('credentials')
@ApiTags('Credentials')
@ApiBearerAuth()
@AllowAuthenticated()
export class CredentialsController {
  constructor(
    private readonly credentialsService: CredentialsService,
    private readonly passwordService: PasswordService,
  ) {}

  @Patch('update-credentials')
  async updateCurrentUserCredentials(
    @GetAuthUser() currentUser: JwtStoredPayload,
    @Body() body: UpdateCredentialDto,
  ) {
    const userId = UUIDFactory.from(currentUser.sub);
    const passwordVo = PasswordFactory.from(body.password);

    const hash = await this.passwordService.hashPassword(passwordVo);

    await this.credentialsService.updateCredentials(userId, hash);

    return new UpdateCredentialResponseDto();
  }
}
