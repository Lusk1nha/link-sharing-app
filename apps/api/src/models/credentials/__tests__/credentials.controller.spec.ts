import { Test, TestingModule } from '@nestjs/testing';
import { CredentialsController } from '../credentials.controller';
import { CredentialsService } from '../credentials.service';
import { UsersModule } from 'src/models/users/users.module';
import { PasswordModule } from 'src/models/password/password.module';
import { ConfigModule } from '@nestjs/config';
import { PrismaService } from 'src/common/database/database.service';
import { RolesModule } from 'src/models/roles/roles.module';
import { SessionsModule } from 'src/models/sessions/sessions.module';
import { CacheModule } from '@nestjs/cache-manager';
import { generateSingleMockUser } from 'src/models/users/__mock__/users.mock';
import { JwtStoredPayload } from 'src/common/auth/__types__/auth.types';
import { Role } from 'src/common/roles/roles.common';
import { UpdateCredentialDto } from '../dto/update-credential.dto';
import { ValidationPipe } from '@nestjs/common';
import { generateSingleMockCredential } from '../__mock__/credentials.mock';
import { CredentialMapper } from '../domain/credential.mapper';
import { UserMapper } from 'src/models/users/domain/user.mapper';
import { UpdateCredentialResponseDto } from '../dto/update-credential-response.dto';
import { PasswordService } from 'src/models/password/password.service';
import { faker } from '@faker-js/faker/.';

describe(CredentialsController.name, () => {
  let controller: CredentialsController;
  let credentialsService: CredentialsService;
  let passwordService: PasswordService;

  beforeEach(() => {
    jest.clearAllMocks(); // zera contadores entre testes
    jest.resetAllMocks(); // remove implementações anteriores
  });

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        RolesModule,
        SessionsModule,
        CacheModule.register({
          isGlobal: true,
        }),
        ConfigModule.forRoot({
          isGlobal: true,
        }),
        PasswordModule,
      ],
      controllers: [CredentialsController],
      providers: [
        {
          provide: CredentialsService,
          useValue: {
            updateCredentials: jest.fn(),
          },
        },
        {
          provide: PasswordService,
          useValue: {
            hashPassword: jest.fn(),
          },
        },
        PrismaService,
      ],
    }).compile();

    controller = module.get<CredentialsController>(CredentialsController);
    credentialsService = module.get<CredentialsService>(CredentialsService);
    passwordService = module.get<PasswordService>(PasswordService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
    expect(credentialsService).toBeDefined();
  });

  describe('updateCurrentUserCredentials', () => {
    it(`should be defined`, () => {
      expect(controller.updateCurrentUserCredentials).toBeDefined();
    });

    it(`should validate the request body`, async () => {
      let target: ValidationPipe = new ValidationPipe({
        transform: true,
        whitelist: true,
      });

      const body: UpdateCredentialDto = {
        password: faker.internet.password({
          pattern: /[A-Za-z0-9!@#$%^&*()_+]/,
        }),
      };

      await expect(
        target.transform(body, {
          type: 'body',
          metatype: UpdateCredentialDto,
        }),
      ).resolves.toEqual(body);
    });

    it(`should validate the request body with invalid data`, async () => {
      let target: ValidationPipe = new ValidationPipe({
        transform: true,
        whitelist: true,
      });

      const body: UpdateCredentialDto = {
        password: '', // Invalid password
      };

      await expect(
        target.transform(body, {
          type: 'body',
          metatype: UpdateCredentialDto,
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it(`should update user credentials`, async () => {
      const mockUser = generateSingleMockUser();
      const mockUserEntity = UserMapper.toDomain(mockUser);

      const body: UpdateCredentialDto = {
        password: 'new-password',
      };

      const mockCredential = generateSingleMockCredential({
        userId: mockUser.id,
        passwordHash: 'new-password-hash',
      });
      const mockCredentialEntity = CredentialMapper.toDomain(mockCredential);

      const jwtUserPayload: JwtStoredPayload = {
        sub: mockUser.id,
        email: mockUser.email,
        roles: [Role.User],
        aud: 'api',
        iat: Date.now(),
        exp: Date.now() + 3600,
        iss: 'test-issuer',
      };

      jest
        .spyOn(passwordService, 'hashPassword')
        .mockResolvedValue('new-password-hash');

      jest
        .spyOn(credentialsService, 'updateCredentials')
        .mockResolvedValue(mockCredentialEntity);

      const response = await controller.updateCurrentUserCredentials(
        jwtUserPayload,
        body,
      );

      expect(credentialsService.updateCredentials).toHaveBeenCalledTimes(1);
      expect(response).toBeInstanceOf(UpdateCredentialResponseDto);
    });
  });
});
