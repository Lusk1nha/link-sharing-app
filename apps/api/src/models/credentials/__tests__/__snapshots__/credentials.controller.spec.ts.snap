// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsController updateCurrentUserCredentials should validate the request body with invalid data 1`] = `"Bad Request Exception"`;
